src/garage/sampler/local_sampler.py

```diff
@@ -31,7 +32,8 @@ def __init__(self, worker_factory, agents, envs):
         # pylint: disable=super-init-not-called
         self._factory = worker_factory
         self._agents = worker_factory.prepare_worker_messages(agents)
-        self._envs = worker_factory.prepare_worker_messages(envs)
+        self._envs = worker_factory.prepare_worker_messages(
+            envs, preprocess=copy.deepcopy)
```

Comment:
seems like this actually might be a bug in the worker factory, no?

@krzentner PTAL

Reply:
I don't think this is a bug in worker factory. Then again, I don't quite understand why this change is necessary either. Only one environment instance gets used at a time, so it should be fine if they're actually all the same environment, right?

Reply:
> Passing EnvUpdate
> 
> ```
>     env = TfEnv(PointEnv())
>     tasks = SetTaskSampler(lambda: TfEnv(PointEnv()))
>     n_workers = 8
>     workers = WorkerFactory(seed=100,
>                             max_path_length=max_path_length,
>                             n_workers=n_workers)
>     sampler = LocalSampler.from_worker_factory(workers, policy, tasks.sample(n_workers))
>     rollouts = sampler.obtain_samples(0,
>                                       161,
>                                       np.asarray(policy.get_param_values()))
> ```

I think this should be the preferred usage, if you can sample *all* of the tasks at the start.
If you can't, you need to use the other one though.

Reply:
The problem is that if we make the worker factory always return separate objects, then the other samplers will end up constructing 2x as many environment instances.

Reply:
I guess one option would be for `prepare_worker_messages` to always return a list of serialized messages (instead of whatever the preprocessing function is). This would break some optimizations in the `RaySampler` though.